File: src/tasks/model/Task.java
package tasks.model;

import omnia.cli.out.Output;
import omnia.data.structure.Set;

public interface Task {

  TaskId id();

  String label();

  Status status();

  boolean isUnblocked();

  Set<? extends Task> blockingTasks();

  Set<? extends Task> blockedTasks();

  Output render();

  enum Status {
    OPEN,
    COMPLETED,
    STARTED;

    public boolean isOpen() {
      return this.equals(OPEN);
    }

    public boolean isCompleted() {
      return this.equals(COMPLETED);
    }

    public boolean isStarted() {
      return this.equals(STARTED);
    }
  }
}

File: engine/src/me/strangepan/tasks/engine/model/Task.kt
package me.strangepan.tasks.engine.model

import omnia.data.structure.Set

/** A read-only immutable model representing the state of a task and its core contents. */
interface Task {
  val store: TaskStore
  val id: TaskId
  val label: String
  val status: Status
  val isUnblocked: Boolean
  val blockingTasks: Set<out Task>
  val blockedTasks: Set<out Task>

  enum class Status {
    OPEN, COMPLETED, STARTED;

    val isOpen: Boolean
      get() = this == OPEN
    val isCompleted: Boolean
      get() = this == COMPLETED
    val isStarted: Boolean
      get() = this == STARTED
  }
}